README.md

```diff
@@ -78,6 +78,95 @@ See also:
 
 ## Usage
 
+### Plugin
+
+**In order to further improve your production efficiency, we have also prepared plug-ins for you**
+
+- provider_template：[Android Studio/Intellij   Plugin](https://plugins.jetbrains.com/plugin/16821-flutter-provider)
+  - one-click generation of necessary folders, files, template codes, etc. for each page
+  - support Wrap Widget: Consumer, Selector, ChangeNotifirProvider
+  - Support input **provider** prompt to generate quick code snippets
+
+### Some usage
+
+#### Usual usage
+
+```dart
+class CounterPage extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return ChangeNotifierProvider(
+      create: (BuildContext context) => CounterProvider(),
+      child: Scaffold(
+        appBar: AppBar(title: Text('Provider')),
+        body: Center(
+          child: Consumer<ProEasyCounterProvider>(
+            builder: (context, provider, child) {
+              return Text('count ${provider.count} ');
+            },
+          ),
+        ),
+        floatingActionButton: FloatingActionButton(
+          onPressed: () {
+            Provider.of<ProEasyCounterProvider>(context, listen: false).increment();
+          },
+          child: Icon(Icons.add),
+        ),
+      ),
+    );
+  }
+}
+
+class CounterProvider extends ChangeNotifier {
+  int count = 0;
+
+  void increment() {
+    count++;
+    notifyListeners();
+  }
+}
+```
+
+#### Some skills
+
+`Provider.of` is a little troublesome to use, we can use some tips to optimize it, ChangeNotifierProvider will store the XxxProvider object, here, we can also store it for subsequent use
+
+```dart
+class CounterPage extends StatelessWidget {
+  final provider = CounterProvider();
+
+  @override
+  Widget build(BuildContext context) {
+    return ChangeNotifierProvider(
+      create: (BuildContext context) => provider,
```

Comment:
This is anti-pattern.

The create function should not reuse an existing variable and instead create a new state.

Reply:
The essence of create function  is to obtain the address of an object, and then store it in _InheritedProviderScopeElement

I think the provider should not be understood as a variable, it represents the address passed to the create function, and then _InheritedProviderScopeElement stores this address, which is the same as directly creating an object.

Because the provider is a final type, there is no need to worry about the provider address being changed

In this way, we can use provider externally instead of Provider.of, which is a bit troublesome.

Reply:
- For those who do not understand the principle of Provider, write the following code, and it will find that Provider.of cannot trigger increase(). This is a wrong use of context, and CounterProvider cannot be found.

```dart
class CounterPage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return ChangeNotifierProvider(
      create: (BuildContext context) => CounterProvider(),
      child: Scaffold(
        appBar: AppBar(title: Text('Provider')),
        body: Center(
          child: Consumer<CounterProvider>(
            builder: (context, provider, child) {
              return Text('count ${provider.count} ');
            },
          ),
        ),
        floatingActionButton: FloatingActionButton(
          onPressed: () {
            Provider.of<CounterProvider>(context, listen: false).increment();
          },
          child: Icon(Icons.add),
        ),
      ),
    );
  }
}

class CounterProvider extends ChangeNotifier {
  int count = 0;

  void increment() {
    count++;
    notifyListeners();
  }
}
```

- If we cache CounterProvider() first and then use it, we can avoid this problem

```dart
class CounterPage extends StatelessWidget {
  final provider = CounterProvider();

  @override
  Widget build(BuildContext context) {
    return ChangeNotifierProvider(
      create: (BuildContext context) => provider,
      child: Scaffold(
        appBar: AppBar(title: Text('Provider')),
        body: Center(
          child: Consumer<CounterProvider>(
            builder: (context, provider, child) {
              return Text('count ${provider.count} ');
            },
          ),
        ),
        floatingActionButton: FloatingActionButton(
          onPressed: () => provider.increment(),
          child: Icon(Icons.add),
        ),
      ),
    );
  }
}

class CounterProvider extends ChangeNotifier {
  int count = 0;

  void increment() {
    count++;
    notifyListeners();
  }
}
```

Reply:
This isn't the correct solution. Instead do:

```dart
ChangeNotifierProvider<Model>(
  create: ...,
  builder: (context, child) {
    return Button(
     onPressed: () => context.read<Model>() // works
   );
  }
)

Reply:
If you don't do that, you can have problems if the widget rebuilds, as you'd lose your reference to the notifier

Reply:
> If you don't do that, you can have problems if the widget rebuilds, as you'd lose your reference to the notifier

create is just a copy of the object we passed in, stored in InheritedElement, I can also store a copy externally.

Reply:
This still is ani-pattern. 

Your application will break if somehow the widget rebuilds.

Reply:
> This still is ani-pattern.
> 
> Your application will break if somehow the widget rebuilds.

your are right